dali/kernels/imgproc/convolution/laplacian_cpu.h

```diff
@@ -0,0 +1,299 @@
+    if (no_smoothing_) {
+      single_dim_impl_->Run(ctx, out, in, windows[deriv_axis], transform);
+    } else {
+      multi_dim_impl_->Run(ctx, out, in, windows, transform);
+    }
+  }
+
+  bool no_smoothing_;
+  std::unique_ptr<MultiDimConv> multi_dim_impl_;
+  std::unique_ptr<SingleDimConv> single_dim_impl_;
+};
+
+
+/**
+ * @brief Provides Laplacian specializations for 2 and 3 dimensional data.
+ *  Run methods expect additional ``acc`` buffer that will be used to accumulate
+ *  partial derivatives. If ``Intermediate`` and ``Out`` are the same type,
+ *  the `acc` and ``out`` can be the same tensor.
+ */
+template <typename T, typename Intermediate, typename Out, typename In, typename W, int axes,
+          bool has_channels>
+struct LaplacianCPUBase;
+
+template <typename T, typename Intermediate, typename Out, typename In, typename W,
+          bool has_channels>
+struct LaplacianCPUBase<T, Intermediate, Out, In, W, 2, has_channels> {
+  static constexpr int axes = 2;
+  using DxKernel = Convolution<Intermediate, In, W, axes, 0, has_channels,
+                               TransScaleSat<Intermediate, W>>;
+  using DyKernel = Convolution<Out, In, W, axes, 1, has_channels, T>;
+  static constexpr int ndim = DyKernel::ndim;  // Dx ndim is the same
+
+  KernelRequirements Setup(KernelContext& ctx, const TensorShape<ndim>& in_shape,
+                           const std::array<std::array<int, axes>, axes>& window_sizes) {
+    KernelRequirements req;
+    req.output_shapes.push_back(uniform_list_shape<ndim>(1, in_shape));
+
+    auto req_dx = sobel_dx_.Setup(ctx, in_shape, window_sizes[0]);
+    auto req_dy = sobel_dy_.Setup(ctx, in_shape, window_sizes[1]);
+
+    // Calculate max scratch memory required by sub-kernels
+    sub_scratch_sizes_ = MaxScratchSize(req_dx.scratch_sizes, req_dy.scratch_sizes);
+    req.scratch_sizes = sub_scratch_sizes_;
+
+    return req;
+  }
+
+  void Run(KernelContext& ctx, const TensorView<StorageCPU, Out, ndim> &out,
+           const TensorView<StorageCPU, Intermediate, ndim> &acc,
+           const TensorView<StorageCPU, const In, ndim>& in,
+           const std::array<std::array<TensorView<StorageCPU, const W, 1>, axes>, axes>& windows,
+           const std::array<float, axes>& scale, const T& transform) {
+    // Prepare the scratchpad with all the remaining memory requested by sub-kernels
+    // TODO(michalz): Get rid of this run-time memory kind dispatch
+    PreallocatedScratchpad sub_scratch;
+    for (size_t i = 0; i < sub_scratch_sizes_.size(); i++) {
+      auto sz = sub_scratch_sizes_[i];
+      auto kind_id = static_cast<mm::memory_kind_id>(i);
+      sub_scratch.allocs[i] = BumpAllocator(
+        static_cast<char*>(ctx.scratchpad->Alloc(kind_id, sz, 64)), sz);
+    }
+
+    KernelContext sub_ctx = ctx;
+    sub_ctx.scratchpad = &sub_scratch;
+
+    // Clear the scratchpad for sub-kernels to reuse memory
+    sobel_dx_.Run(sub_ctx, acc, in, windows[0], scale[0]);
```

Comment:
```suggestion
    sobel_dx_.Run(sub_ctx, acc, in, windows[0], {scale[0]});
```
Shouldn't we create a transform here? How does it work with a scale? Is it implicit conversion or something?

Reply:
Also, we pass 2 scales (axes = 2) and only use the one here. Is it intended?

Same in the 3D case.

Reply:
Oh, I see it's packed into the separate transform.